Sources/Real/Field.swift

```diff
@@ -0,0 +1,55 @@
+//===--- Field.swift ------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Numerics open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift Numerics project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+// TODO: refining Numeric is slightly annoying because Numeric has the
+// magnitude property and associated type, which doesn't quite make sense
+// for all fields. However, simply picking magnitude = self in those cases
+// is probably a sensible workaround, so this isn't catastrophic. Still,
+// slightly annoying.
+
+/// A type modeling an algebraic field.
+///
+/// This protocol refines `Numeric`, adding division and reciprocal operations.
+public protocol Field: Numeric {
+  
+  static func /=(a: inout Self, b: Self)
+  
+  static func /(a: Self, b: Self) -> Self
+  
+  /// The (approximate) reciprocal (multiplicative inverse) of this number, if it is representable.
+  ///
+  /// If reciprocal is non-nil, you can replace division by self with multiplication by reciprocal and
+  /// either get exact the same result (for finite fields) or approximately the same result up to a
+  /// typical rounding error (for floating-point formats).
+  ///
+  /// If self is zero, or if a reciprocal would overflow or underflow such that it cannot be accurately
+  /// represented, the result is nil. Implementations should be *conservative*; it is OK to return
+  /// nil even in some cases where a reciprocal can be represented. For this reason, a default
+  /// implementation that simply always returns nil is provided.
+  ///
+  /// Note that `.zero.reciprocal`, somewhat surprisingly, is *not* nil for `Real` or
+  /// `Complex` types, because these types have an `.infinity` value that acts as the
+  /// reciprocal of `.zero`.
+  var reciprocal: Self? { get }
```

Comment:
What is the rationale for making the `reciprocal` property `Optional`?

For comparison, the unary-negation operator in the standard library is non-optional and traps on overflow (eg. `-Int.min`).

Reply:
The conditions under which unary-negation would trap are trivially detectable *a priori*. The conditions under which a reciprocal would deliver a result that is either incorrect or has significantly reduced accuracy are significantly harder to characterize, and users are likely to either fail to detect them, or get them wrong. Returning an optional lets us handle the check for them.

Reply:
(This glosses over the fact that detecting when unary-negation would trap _in a protocol context_ is non-trivial for all the named protocols in the hierarchy under SignedNumeric, so we probably *should* add an optional `negation` property at some point.)

Reply:
That’s actually really clever. People who want the trapping behavior can write `1/x`, and people who want to check for failure can use `reciprocal`.

+1 to this design, and +1 to bringing it to the standard library for `negated` as well.